### PR TITLE
Implement frontend throttle

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
Implementing the FE throttle

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-services/issues/8355

**Notes for reviewer**:
* Branching off of the BE throttle story
* Reusing the throttle login from it
* Wrapping it into a HapiThrottle class and using _synchronize_ for the methods
* Deleting not needed classes

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/GeneralThrottleAccumulator.java [hunk 1]
```diff
@@ -106,6 +109,44 @@ public boolean shouldThrottle(@NonNull TransactionInfo txnInfo, Instant t, Heder
...
+                configuration.getConfigData(ContractsConfig.class).throttleThrottleByGas();
+
+        // Note that by payer exempt from throttling we mean just that those transactions will not be throttled,
+        // such payer accounts neither impact the throttles nor are they impacted by them
+        // In the current mono-service implementation we have the same behavior, additionally it is
+        // possible that transaction can also be exempt from affecting congestion levels separate from throttle
+        // exemption
+        // but this is only possible for the case of triggered transactions which is not yet implemented (see
+        // MonoMultiplierSources.java)
+        final var payer = PbjConverter.toPbj(query.getAccountDetails().getAccountId());
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/GeneralThrottleAccumulator.java:124]: Is this a correct way to get the payer?

Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/GeneralThrottleAccumulator.java:124]: We are in the modularized code-base and should not use old protobuf objects in the method signatures. We should use PBJ-objects as long as possible and only convert right before and after we call mono-code.

The parameter `query` has to have the full name `com.hedera.hapi.node.transaction.Query`. When you do that, `query.accountDetails().account()` will have the right type.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java [hunk 2]
```diff
@@ -147,7 +147,9 @@ public void handleQuery(@NonNull final Bytes requestBuffer, @NonNull final Buffe
             }
 
             // 3. Check query throttles
-            if (throttleAccumulator.shouldThrottleQuery(function, query)) {
+            if (hapiThrottling.shouldThrottleQuery(
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java:150]: As stated above, we should not convert to proto here.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java [hunk 3]
```diff
@@ -16,13 +16,7 @@
...
 import dagger.Module;
-import javax.inject.Singleton;
 
 @Module
-public interface ThrottleInjectionModule {
-    @Binds
-    @Singleton
-    ThrottleAccumulator bindThrottleAccumulator(ThrottleAccumulatorImpl throttleAccumulator);
-}
+public interface ThrottleInjectionModule {}
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java:22]: Should I delete the whole class or leave it for future additions?

Comment (tinker-michaelj) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleInjectionModule.java:22]: Delete it imo. 😁

**Review by tinker-michaelj — COMMENTED** [review]
I can't immediately see where the frontend throttle's capacity is being "split" by the number of nodes in the network. Where does that happen?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/GeneralThrottleAccumulator.java [hunk 4]
```diff
@@ -90,7 +91,7 @@ public class HandleThrottleAccumulator {
     private GasLimitDeterministicThrottle gasThrottle;
     private List<DeterministicThrottle> activeThrottles = Collections.emptyList();
 
-    public HandleThrottleAccumulator(@NonNull final ConfigProvider configProvider) {
+    public GeneralThrottleAccumulator(@NonNull final ConfigProvider configProvider) {
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/GeneralThrottleAccumulator.java:94]: I think we should name these classes differently. `GeneralThrottleAccumulator` and `HapiThrottling` provide almost the same functionality, but the names follow different patterns.

What do you think about calling this class `ThrottleAccumulator` and renaming `HapiThrottling` to `SynchronizedThrottleAccumulator`?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java [hunk 5]
```diff
@@ -49,19 +49,19 @@
...
 /**
- * Implementation of {@link NetworkUtilizationManager}  that delegates to injected {@link HandleThrottleAccumulator} and {@link
+ * Implementation of {@link NetworkUtilizationManager}  that delegates to injected {@link GeneralThrottleAccumulator} and {@link
  * MultiplierSources}.
  */
 public class NetworkUtilizationManagerImpl implements NetworkUtilizationManager {
     private static final Logger log = LogManager.getLogger(NetworkUtilizationManagerImpl.class);
 
-    private final HandleThrottleAccumulator handleThrottling;
+    private final GeneralThrottleAccumulator handleThrottling;
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/impl/NetworkUtilizationManagerImpl.java:58]: This property should be renamed to something closer to the class name IMO.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HapiThrottling.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    public HapiThrottling(GeneralThrottleAccumulator generalThrottleAccumulator) {
+        this.generalThrottleAccumulator =
+                requireNonNull(generalThrottleAccumulator, "generalThrottleAccumulator must not be null");
+    }
+
+    public synchronized boolean shouldThrottle(@NonNull TransactionInfo txnInfo, HederaState state) {
+        return generalThrottleAccumulator.shouldThrottle(txnInfo, Instant.now(), state);
+    }
+
+    public synchronized boolean shouldThrottleQuery(Query query, HederaFunctionality queryFunction) {
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HapiThrottling.java:43]: Keeping the order of parameters consistent makes it easier to learn and use an API. In general, we are not consistent in our API, and one day, I want us to define an order and go through the API to clean this up.

But in this particular case, we should make the methods consistent right away because `GeneneralThrottleAccumulator.shouldThrottleQuery()` and `HappiThrottling.shouldThrottleQuery()` are so similar. `HederaFunctionality` always comes first, and `Query` last or vice versa.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java [hunk 7]
```diff
@@ -171,7 +171,8 @@ public void handleQuery(@NonNull final Bytes requestBuffer, @NonNull final Buffe
...
+                    query, HederaFunctionality.fromProtobufOrdinal(function.protoOrdinal()))) {
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java:175]: Can we use `function`?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HapiThrottling.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    private final GeneralThrottleAccumulator generalThrottleAccumulator;
+
+    @Inject
+    public HapiThrottling(GeneralThrottleAccumulator generalThrottleAccumulator) {
+        this.generalThrottleAccumulator =
+                requireNonNull(generalThrottleAccumulator, "generalThrottleAccumulator must not be null");
+    }
+
+    public synchronized boolean shouldThrottle(@NonNull TransactionInfo txnInfo, HederaState state) {
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/HapiThrottling.java:39]: Sorry, I did not realize this when reviewing `HandleThrottleAccumulator`. I think we should make the names of the methods consistent. Either both methods are called `shouldThrotte()` or we call them `shouldThrottleTransaction()` and `shouldThrottleQuery()`. I'd prefer the first but have no strong opinion.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/throttle/HapiThrottlingTest.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    @Test
+    void verifyShouldThrottleIsCalled() {
+        // given
+        final var state = mock(HederaState.class);
+
+        // when
+        subject.shouldThrottle(transactionInfo, state);
+
+        // then
+        verify(generalThrottleAccumulator, times(1)).shouldThrottle(eq(transactionInfo), any(), eq(state));
```
Comment (netopyr) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/throttle/HapiThrottlingTest.java:60]: I believe `times(1)` is the default and unnecessary here.

**Review by netopyr — CHANGES_REQUESTED** [review]
I agree with @tinker-michaelj. It is not immediately clear where the capacity is split by the number of nodes.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 10]
```diff
@@ -16,35 +16,550 @@
...
 /**
- * Keeps track of the amount of usage of different throttle categories (by {@code id}), and returns
- * whether the throttle has been exceeded after applying the given incremental amount.
+ * Keeps track of the amount of usage of different TPS throttle categories and gas, and returns whether a given
+ * transaction or query should be throttled based on that.
+ * Meant to be used in single-threaded context only as part of the {@link com.hedera.node.app.workflows.handle.HandleWorkflow}.
  */
-public interface ThrottleAccumulator {
+@Singleton
+public class ThrottleAccumulator {
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:81]: We had `ThrottleAccumulator` interface that was not needed and I deleted it. Then I renamed HandleThrottleAccumulator to ThrottleAccumulator and now GIT is showing that I've deleted HandleThrottleAccumulator and changed ThrottleAccumulator. In reality the only change for this class(HandleThrottleAccumulator) will be the rename and addition of CAPACITY_SPLIT(not implemented yet, will do that today)

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java [hunk 11]
```diff
@@ -804,9 +805,10 @@ private void initializeDagger(@NonNull final MerkleHederaState state, @NonNull f
                             throttleManager,
                             exchangeRateManager,
                             monoMultiplierSources,
-                            handleThrottling))
+                            throttleAccumulator))
                     .networkUtilizationManager(
-                            new NetworkUtilizationManagerImpl(handleThrottling, monoMultiplierSources))
+                            new NetworkUtilizationManagerImpl(throttleAccumulator, monoMultiplierSources))
+                    .synchronizedThrottleAccumulator(new SynchronizedThrottleAccumulator(throttleAccumulator))
```
Comment (MiroslavGatsanoga) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:811]: I think we need a separate instance of `ThrottleAccumulator` for the synchronized one with capacity split equal to the number of nodes (i.e. the FE throttling), leaving the current `throttleAccumulator` with capacity split of 1 to be the one used in the handle workflow and congestion (i.e. the BE throttling).
You can check [the mono setup](https://github.com/hashgraph/hedera-services/blob/25bf841331b46c748a081b0f01a1ad1e86cb21e4/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/throttling/ThrottlingModule.java#L33) for reference.

P.S. Having separate instances for FE and BE throttle would mean we need to call `rebuildFor` and `applyGasConfig` methods for both.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/throttle/HapiThrottlingTest.java (hunk 9 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/throttle/HapiThrottlingTest.java]: Yes, the default is `times(1)` but writing it explicitly seems more clear to me. Looking at the code we have a fair amount of asserts that are written that way. What do you think?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java [hunk 12]
```diff
@@ -171,7 +172,8 @@ public final class Hedera implements SwirldMain {
      */
     private PlatformStatus platformStatus = PlatformStatus.STARTING_UP;
 
-    private HandleThrottleAccumulator handleThrottling;
+    private ThrottleAccumulator throttleAccumulator;
+    private ThrottleAccumulator synchronizedThrottleAccumulator;
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:176]: Any better ideas for naming this field? Since we have a `SynchronizedThrottleAccumulator` class and it's a bit confusing. Maybe `throttleAccumulatorWithCapacitySplit`?

**Comment by vtronkov** [comment]
There are a lot of failing tests due to the FE throttle. I will investigate

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java (hunk 12 above)
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:176]: We could use `hapiThrottling` and `handleThrottling` maybe.

Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:176]: Or `backendThrottle` and `frontendThrottle`. If we use these terms consistently throughout the code base, they will become project language.

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java [hunk 13]
```diff
@@ -48,5 +48,5 @@ public record BootstrapConfig(
         @ConfigProperty(value = "rates.nextExpiry", defaultValue = "4102444800") @NetworkProperty long ratesNextExpiry,
         @ConfigProperty(value = "system.entityExpiry", defaultValue = "1812637686") @NetworkProperty
                 long systemEntityExpiry,
-        @ConfigProperty(value = "throttleDefsJson.resource", defaultValue = "throttles.json") @NodeProperty
+        @ConfigProperty(value = "throttleDefsJson.resource", defaultValue = "throttles-dev.json") @NodeProperty
```
Comment (vtronkov) [inline hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java:51]: Looks like the mono is using the `throttles-dev.json` instead of `throttles.json`. `UmbrellaRedux.umbrellaRedux` test was failing due to configuration differences - mainly because of the `CryptoCreate`. In `throttles-dev.json` the _milliOpsPerSec_ value is set to **2000000**, in `throttles.json` to **2000**.

@MiroslavGatsanoga suggested another option - change the test to load the correct config with 
`protoDefsFromResource("testSystemFiles/throttles-dev.json")`  - with that change the test is also passing.

@tinker-michaelj @netopyr thoughts?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 14]
```diff
@@ -16,35 +16,553 @@
...
+    private EnumMap<HederaFunctionality, ThrottleReqsManager> functionReqs = new EnumMap<>(HederaFunctionality.class);
+    private static final int UNKNOWN_NUM_IMPLICIT_CREATIONS = -1;
+    private boolean lastTxnWasGasThrottled;
+    private GasLimitDeterministicThrottle gasThrottle;
+    private List<DeterministicThrottle> activeThrottles = Collections.emptyList();
+
+    public ThrottleAccumulator(
+            @NonNull final IntSupplier capacitySplitSource, @NonNull final ConfigProvider configProvider) {
+        this.configProvider = requireNonNull(configProvider, "configProvider must not be null");
+        this.capacitySplit = capacitySplitSource.getAsInt();
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:98]: Why do you use an `IntSupplier` instead of passing the `int` directly?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java [hunk 15]
```diff
@@ -171,7 +173,11 @@ public void handleQuery(@NonNull final Bytes requestBuffer, @NonNull final Buffe
             }
 
             // 3. Check query throttles
-            if (throttleAccumulator.shouldThrottleQuery(function, query)) {
+            if (synchronizedThrottleAccumulator.shouldThrottle(function, query)
+                    && !RESTRICTED_FUNCTIONALITIES.contains(function)) {
+                if (!responseType.equals(COST_ANSWER)) {
+                    throw new PreCheckException(NOT_SUPPORTED);
```
Comment (netopyr) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java:179]: This comes as a surprise. Why are we returning `NOT_SUPPORTED` here?

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java (hunk 13 above)
Comment (netopyr) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java:51]: What are we using in production currently? The default value of any `ConfigProperty` should reflect that.

File: hedera-node/hedera-config/src/testFixtures/java/com/hedera/node/config/testfixtures/HederaTestConfigBuilder.java [hunk 16]
```diff
@@ -175,6 +175,7 @@ public static TestConfigBuilder create() {
                 .withConfigDataType(StakingConfig.class)
                 .withConfigDataType(StatsConfig.class)
                 .withConfigDataType(TokensConfig.class)
+                .withConverter(new ScaleFactorConverter())
```
Comment (netopyr) [inline hedera-node/hedera-config/src/testFixtures/java/com/hedera/node/config/testfixtures/HederaTestConfigBuilder.java:178]: Is this change necessary, or can we revert it?

Comment (vtronkov) [reply hedera-node/hedera-config/src/testFixtures/java/com/hedera/node/config/testfixtures/HederaTestConfigBuilder.java]: Reverted, I guess something happened when merging with master

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java (hunk 14 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:98]: Just to keep it consistent with the mono implementation(DeterministicThrottling). I can change it with **int** if you think it's better.

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java (hunk 13 above)
Comment (vtronkov) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java:51]: I'm not sure exactly which of all configs are for production but I doubt that the value is `throttles-dev.json` 😄 I can revert this change but is it okay to change the suite to use the dev throttle file?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java (hunk 14 above)
Comment (MiroslavGatsanoga) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:98]: I think the idea is when we initialize the frontend throttle like this
```
this.frontendThrottle =
                new ThrottleAccumulator(() -> platform.getAddressBook().getSize(), configProvider);
```
is to be able to get the address book size dynamically and use it for the capacity split but here we get the capacity split value only in the constructor which seems like something we should fix actually.

In mono we use the `IntSupplier` in [rebuildFor()](https://github.com/hashgraph/hedera-services/blob/06b8d8ae123ece1ab72c35f43c1097d1dfb0dc9b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/throttling/DeterministicThrottling.java#L197) so it should pick up the address books size dynamically.

Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:98]: That makes sense! I will fix that. Thanks

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java (hunk 15 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java:179]: `UnsupportedQueriesRegression.verifyUnsupportedOps` is failing with `expected: <NOT_SUPPORTED> but was: <BUSY>!`

Debugging the mono implementation - the throttle is returning back **BUSY** response code then we go through `GetLiveHashAnswer.responseGiven` and we return **NOT_SUPPORTED** instead of **BUSY** if `type != COST_ANSWER`. But now I notice that we are returning **NOT_SUPPORTED** if the `type == COST_ANSWER` as well 🤔 

This is happening when we call **hygieneCheck** in `StakedAnswerFlow.satisfyUsing` and if the `hygieneCheck != OK`. Maybe this is some kind of a pre-check and I should move that logic there? Not sure, any idea? I will keep looking at that.

Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java]: Okay, I deleted my previous comment since it's confusing.

I think I understand the problem now. The immediate fix I can think of is to move the throttling after the validations. You guys can tell me if this is a problem but I can't think of a different solution.

If the throttling is before the validation - in the `shouldThrottle` method we try to get the ThrottleReqsManager but since these operations(listed in `UnsupportedQueriesRegression.verifyUnsupportedOps`) are not supported we don't have _managers_ and we return BUSY.

Any concerns about this change?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java (hunk 14 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java]: @MiroslavGatsanoga fixed

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java (hunk 15 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/query/QueryWorkflowImpl.java]: The other option would be to throw `NOT_SUPPORTED` inside the `shouldThrottle` method if the _manager_ is not found.

File: hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java (hunk 13 above)
Comment (netopyr) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java:51]: I just checked the default in `BootstrapConfig` has to be `throttles.json`.

How do you see that the tests in mono are using `throttles-dev.json`? I think it makes sense to update the test and make the usage of `throttles-dev.json` more explicit.

Comment (vtronkov) [reply hedera-node/hedera-config/src/main/java/com/hedera/node/config/data/BootstrapConfig.java:51]: Thanks, Miro. I wasn't sure exactly which property file was using mono but when I debugged it I saw that the throttle values are as `throttles-dev.json`.

I will revert the BootstrapConfig change and I will update the test 👍

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallLocalSuite.java [hunk 17]
```diff
@@ -285,7 +285,6 @@ private HapiSpec erc20Query() {
     }
 
     // https://github.com/hashgraph/hedera-services/pull/5485
-    @HapiTest
     private HapiSpec callLocalDoesNotCheckSignaturesNorPayer() {
```
Comment (vtronkov) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallLocalSuite.java:288]: Disabling this test because it's failing because it's failing in mono as well. My theory is that it was succeeding until now because the throttles were not implemented.

@MiroslavGatsanoga noticed that this test is part of `ConcurrentSuites` and it uses `throttles-dev.json` and it's running on a single node - I tried that but it's still failing.

**Review by netopyr — COMMENTED** [review]
LGTM but @tinker-michaelj should also review, just to be sure.

**Review by netopyr — APPROVED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 18]
```diff
@@ -16,35 +16,553 @@
...
+        // possible that transaction can also be exempt from affecting congestion levels separate from throttle
+        // exemption
+        // but this is only possible for the case of triggered transactions which is not yet implemented (see
+        // MonoMultiplierSources.java)
+        final var isPayerThrottleExempt = throttleExempt(txnInfo.payerID(), configuration);
+        if (isPayerThrottleExempt) {
+            return false;
+        }
+
+        // TODO: in mono we pass: () -> getSpanMapAccessor().getEthTxDataMeta(this) as 3rd param to
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:235]: That's a performance optimization; the behavior here should be functionally identical.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 19]
```diff
@@ -16,35 +16,553 @@
...
 
 /**
- * Keeps track of the amount of usage of different throttle categories (by {@code id}), and returns
- * whether the throttle has been exceeded after applying the given incremental amount.
+ * Keeps track of the amount of usage of different TPS throttle categories and gas, and returns whether a given
+ * transaction or query should be throttled based on that.
+ * Meant to be used in single-threaded context only as part of the {@link com.hedera.node.app.workflows.handle.HandleWorkflow}.
  */
-public interface ThrottleAccumulator {
+@Singleton
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:81]: Since we aren't managing its instances with Dagger2, this isn't technically a but; but since this type isn't a `@Singleton` imo we should remove this annotation.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 20]
```diff
@@ -16,35 +16,553 @@
...
+        if (contractsConfig.throttleThrottleByGas() && contractsConfig.maxGasPerSec() == 0) {
+            log.warn("Consensus gas throttling enabled, but limited to 0 gas/sec");
+            return;
+        }
+
+        final long capacity = contractsConfig.maxGasPerSec();
+        gasThrottle = new GasLimitDeterministicThrottle(capacity);
+
+        log.info(
+                "Resolved consensus gas throttle -\n {} gas/sec (throttling {})",
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:530]: ⚠️ Hard-coding "consensus" here will result in some confusing logs, since this class also implements the frontend throttle. Imo we should make the type of throttle a constructor parameter.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 21]
```diff
@@ -16,35 +16,553 @@
...
+    }
+
+    @NonNull
+    private ThrottleGroup<HederaFunctionality> hapiGroupFromPbj(
+            @NonNull final com.hedera.hapi.node.transaction.ThrottleGroup pbjThrottleGroup) {
+        return new ThrottleGroup<>(pbjThrottleGroup.milliOpsPerSec(), pbjThrottleGroup.operations());
+    }
+
+    private void logResolvedDefinitions(final int capacitySplit) {
+        var sb = new StringBuilder("Resolved handle throttles (after splitting capacity ")
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:542]: Same here---for clarity we'd prefer to log "frontend throttles" as appropriate.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java [hunk 22]
```diff
@@ -16,35 +16,553 @@
...
+
+    /*
+     * Updates the throttle requirements for the given query and returns whether the query should be throttled.
+     *
+     * @param queryFunction the functionality of the query
+     * @param now the time at which the query is being processed
+     * @param query the query to update the throttle requirements for
+     * @return whether the query should be throttled
+     */
+    public boolean shouldThrottle(
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java:131]: ⚠️ I think we should add a parameter `AccountID queryPayerId` to this method's arguments and then do the same `isPayerThrottleExempt()` check you do [here in the transaction throttling logic](https://github.com/hashgraph/hedera-services/blob/08355-implement-frontend-throttle/hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java#L230); this should allow the `callLocalDoesNotCheckSignaturesNorPayer()` spec below to continue passing.

**Review by tinker-michaelj — APPROVED** [review]
I have a few concerns but couldn't bear to make you re-run all checks haha 😁---nice work! Approved here and created [this issue](https://github.com/hashgraph/hedera-services/issues/9244) for follow-ups
